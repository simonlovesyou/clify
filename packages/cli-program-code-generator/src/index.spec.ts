import type { Schema as CommandLineSchema } from "@cling/parser";
import createCli, { ParsedProgram } from ".";

const DEFAULT_PROGRAM_PARAMS = {
  async: false,
  exported: true,
  filePath: "foo.ts",
  defaultExport: true,
  internal: false,
  public: true,
  description: "String",
};

describe("single function", () => {
  describe("two required params", () => {
    const parsedProgram: ParsedProgram = [
      {
        name: "add",
        schema: {
          type: "object",
          properties: {
            numberA: {
              type: "number",
            },
            numberB: {
              type: "number",
            },
          },
          required: ["numberA", "numberB"],
        },
        ...DEFAULT_PROGRAM_PARAMS,
      },
    ];
    describe("options.positionals = false", () => {
      const options = {
        positionals: false,
      };
      it("should output the correct program", () => {
        expect(createCli(parsedProgram, options)).toMatchSnapshot();
      });
    });
    describe("options.positionals = true", () => {
      const options = {
        positionals: true,
      };
      it("should output the correct program", () => {
        expect(createCli(parsedProgram, options)).toMatchSnapshot();
      });
    });
  });
  describe("two optional params", () => {
    const parsedProgram: ParsedProgram = [
      {
        name: "add",
        schema: {
          type: "object",
          properties: {
            numberA: {
              type: "number",
            },
            numberB: {
              type: "number",
            },
          },
        },
        ...DEFAULT_PROGRAM_PARAMS,
      },
    ];
    it("should output the correct program", () => {
      expect(createCli(parsedProgram, {})).toMatchSnapshot();
    });
  });
  describe("one required param", () => {
    describe('one optional param', () => {
      const parsedProgram: ParsedProgram = [
        {
          name: "add",
          schema: {
            type: "object",
            properties: {
              numberA: {
                type: "number",
              },
              numberB: {
                type: 'number'
              }
            },
            required: ['numberA']
          },
          ...DEFAULT_PROGRAM_PARAMS,
        },
      ];
      it("should output the correct program", () => {
        expect(createCli(parsedProgram, {})).toMatchSnapshot();
      });
    })
    describe('one optional param in object', () => {
      const parsedProgram: ParsedProgram = [
        {
          name: "add",
          schema: {
            type: "object",
            properties: {
              numberA: {
                type: "number",
              },
              options: {
                type: 'object',
                properties: {
                  numberB: {
                    type: 'number'
                  }
                }
              }
            },
            required: ['numberA']
          },
          ...DEFAULT_PROGRAM_PARAMS,
        },
      ];
      it("should output the correct program", () => {
        expect(createCli(parsedProgram, {})).toMatchSnapshot();
      });
    })
  });
});
