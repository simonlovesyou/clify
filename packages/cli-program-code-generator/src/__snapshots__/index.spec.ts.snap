// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`single function one required param one optional param in object should output the correct program 1`] = `
"#!/usr/bin/env node
import parser from \\"@cling/parser\\";
import { default as add } from \\"./foo\\";

try {
  const programSchema = {
    arguments: {
      numberA: {
        type: \\"number\\" as const,
      },
    },
    options: {
      numberB: {
        type: \\"number\\" as const,
      },
    },
  };

  const { arguments, options } = parser(programSchema);

  const result = add.apply(null, [
    arguments[\\"numberA\\"],
    {
      [\\"numberB\\"]: options[\\"numberB\\"],
    },
  ]);

  console.log(result);
  return process.exit(0);
} catch (error) {
  console.error(error.message);
  process.exit(1);
}
"
`;

exports[`single function one required param one optional param should output the correct program 1`] = `
"#!/usr/bin/env node
import parser from \\"@cling/parser\\";
import { default as add } from \\"./foo\\";

try {
  const programSchema = {
    arguments: {
      numberA: {
        type: \\"number\\" as const,
      },
    },
    options: {
      numberB: {
        type: \\"number\\" as const,
      },
    },
  };

  const { arguments, options } = parser(programSchema);

  const result = add.apply(null, [arguments[\\"numberA\\"], options[\\"numberB\\"]]);

  console.log(result);
  return process.exit(0);
} catch (error) {
  console.error(error.message);
  process.exit(1);
}
"
`;

exports[`single function two optional params should output the correct program 1`] = `
"#!/usr/bin/env node
import parser from \\"@cling/parser\\";
import { default as add } from \\"./foo\\";

try {
  const programSchema = {
    options: {
      numberA: {
        type: \\"number\\" as const,
      },
      numberB: {
        type: \\"number\\" as const,
      },
    },
  };

  const { options } = parser(programSchema);

  const result = add.apply(null, [options[\\"numberA\\"], options[\\"numberB\\"]]);

  console.log(result);
  return process.exit(0);
} catch (error) {
  console.error(error.message);
  process.exit(1);
}
"
`;

exports[`single function two required params options.positionals = false should output the correct program 1`] = `
"#!/usr/bin/env node
import parser from \\"@cling/parser\\";
import { default as add } from \\"./foo\\";

try {
  const programSchema = {
    arguments: {
      numberA: {
        type: \\"number\\" as const,
      },
      numberB: {
        type: \\"number\\" as const,
      },
    },
  };

  const { arguments } = parser(programSchema);

  const result = add.apply(null, [arguments[\\"numberA\\"], arguments[\\"numberB\\"]]);

  console.log(result);
  return process.exit(0);
} catch (error) {
  console.error(error.message);
  process.exit(1);
}
"
`;

exports[`single function two required params options.positionals = true should output the correct program 1`] = `
"#!/usr/bin/env node
import parser from \\"@cling/parser\\";
import { default as add } from \\"./foo\\";

try {
  const programSchema = {
    positionals: [
      {
        type: \\"number\\" as const,
      },
      {
        type: \\"number\\" as const,
      },
    ],
  };

  const { positionals } = parser(programSchema);

  const result = add.apply(null, [positionals[\\"0\\"], positionals[\\"1\\"]]);

  console.log(result);
  return process.exit(0);
} catch (error) {
  console.error(error.message);
  process.exit(1);
}
"
`;
